### PR TITLE
BUG: fix optimize.dual_annealing does not pass arguments to jacobian

### PR DESCRIPTION
**Pull request by sgubianpm**
See #11052
I also added unit test to check that args are passed to ```jac``` when jacobian is provided.

### CODE REVIEW

File: scipy/optimize/_dual_annealing.py [hunk 1]
```diff
@@ -402,6 +410,10 @@ def __init__(self, search_bounds, func_wrapper, **kwargs):
                 'maxiter': ls_max_iter,
             }
             self.kwargs['bounds'] = list(zip(self.lower, self.upper))
+        if 'jac' in self.kwargs:
+            # Make sure jac wrapper is used so that args are passed.
+            func_wrapper.jac = self.kwargs['jac']
```
Comment (andyfaff) [inline scipy/optimize/_dual_annealing.py:415]: I'm not sure this alteration will work. From the `minimize` [documentation for jac](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize), `jac` can be one of `{callable, ‘2-point’, ‘3-point’, ‘cs’, bool}`. So the `minimizer_kwargs['jac']` is not necessarily a callable.

Comment (sgubianpm) [reply scipy/optimize/_dual_annealing.py:415]: Good point! Thanks @andyfaff, then, we would need this only when ```jac``` is a callable, and if it is anything else, the ```func_wrapper.jac``` is not used and the ```jac``` option is passed as-is to the local minimizer.

**Review by mdhaber — COMMENTED** [review]
It would be good to fix gh-11052, so I merged main here, but this branch won't let me push. Could you [Allow Edits from Maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests), @sgubianpm?

Also, why is the code split between `ObjectiveFunWrapper` and `LocalSearchWrapper` in this way? On one hand, it doesn't look like `ObjectiveFunWrapper` really needs to be involved at all - the Jacobian function could be wrapped completely in `LocalSearchWrapper.__init__`. On the other hand, if you want `ObjectiveFunWrapper` to "know" something about its Jacobian ("gradient", really, right?) then why not pass that information into `ObjectiveFunWrapper.__init__` rather than setting the attribute `func_wrapper.jac = self.kwargs['jac']` from `LocalSearchWrapper.__init__`?

Thanks!

**Comment by mdhaber** [comment]
Could you [Allow Edits from Maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests), @sgubianpm?

**Comment by mdhaber** [comment]
Could you [Allow Edits from Maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) , @sgubianpm?

**Comment by sgubianpm** [comment]
> It would be good to fix [gh-11052](https://github.com/scipy/scipy/issues/11052), so I merged main here, but this branch won't let me push. Could you [Allow Edits from Maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests), @sgubianpm?
> 
> Also, why is the code split between `ObjectiveFunWrapper` and `LocalSearchWrapper` in this way? On one hand, it doesn't look like `ObjectiveFunWrapper` really needs to be involved at all - the Jacobian function could be wrapped completely in `LocalSearchWrapper.__init__`. On the other hand, if you want `ObjectiveFunWrapper` to "know" something about its Jacobian ("gradient", really, right?) then why not pass that information into `ObjectiveFunWrapper.__init__` rather than setting the attribute `func_wrapper.jac = self.kwargs['jac']` from `LocalSearchWrapper.__init__`?
> 
> Thanks!

Did not see that one. I think that this has to be fixed indeed. I can change the PR according your comment...

**Comment by mdhaber** [comment]
@sgubianpm Thanks. If you'd like me to push my commit that merge main, please let me know when you've allowed edits by maintainers. I'd appreciate it if you could also take a look at gh-10437.

**Comment by mdhaber** [comment]
Since there has been no followup, I'll go ahead and close this in favor of gh-17956. Please consider reviewing that or looking into gh-10437.

**Pull request closed without merging**